libyul/backends/evm/ssa/Stack.h

```diff
@@ -0,0 +1,309 @@
+
+template<typename StackManipulationCallback>
+concept StackManipulationCallbackConcept = requires(
+	StackManipulationCallback& _callback,
+	StackSlot _slot,
+	size_t _depth
+)
+{
+	{ _callback.swap(_depth) } -> std::same_as<void>;
+	{ _callback.dup(_depth) } -> std::same_as<void>;
+	{ _callback.push(_slot) } -> std::same_as<void>;
+	{ _callback.pop() } -> std::same_as<void>;
+};
+
+struct NoOpStackManipulationCallbacks
+{
+	static void swap(size_t) {}
+	static void dup(size_t) {}
+	static void push(StackSlot const&) {}
+	static void pop() {}
+};
+static_assert(StackManipulationCallbackConcept<NoOpStackManipulationCallbacks>);
+
+template<
+	StackManipulationCallbackConcept CallbacksType = NoOpStackManipulationCallbacks
+>
+class Stack
+{
+	static size_t constexpr reachableStackDepth = 16;
+public:
+	using Callbacks = CallbacksType;
+
+	using Slot = StackSlot;
+	using Data = StackData;
+
+	/// Array index into stack from the bottom (offset 0 = bottom).
+	/// Natural for array-like access and iteration; used when treating the stack as a data structure.
+	struct Offset
+	{
+		explicit constexpr Offset(size_t _value) : value(_value) {}
+		size_t value;
+		auto operator<=>(Offset const&) const = default;
+	};
+	// comparison operations with size_t
+	friend constexpr auto operator<=>(Offset lhs, size_t rhs) noexcept { return lhs.value <=> rhs; }
+	friend constexpr auto operator<=>(size_t lhs, Offset rhs) noexcept { return lhs <=> rhs.value; }
+	friend constexpr bool operator==(Offset lhs, size_t rhs) noexcept { return lhs.value == rhs; }
+	friend constexpr bool operator==(size_t lhs, Offset rhs) noexcept { return lhs == rhs.value; }
+
+	/// Distance from the stack top (depth 0 = top).
+	/// Natural for stack operations (SWAP1 = swap with depth 1); used for operations that
+	/// conceptually work "from the top".
+	struct Depth
+	{
+		explicit constexpr Depth(size_t _value) : value(_value) {}
+		size_t value;
+		auto operator<=>(Depth const&) const = default;
+	};
+	// comparison operations with size_t
+	friend constexpr auto operator<=>(Depth lhs, size_t rhs) noexcept { return lhs.value <=> rhs; }
+	friend constexpr auto operator<=>(size_t lhs, Depth rhs) noexcept { return lhs <=> rhs.value; }
+	friend constexpr bool operator==(Depth lhs, size_t rhs) noexcept { return lhs.value == rhs; }
+	friend constexpr bool operator==(size_t lhs, Depth rhs) noexcept { return lhs == rhs.value; }
+
+	Stack(
+		Data& _data,
+		Callbacks _callbacks
+	):
+		m_data(&_data),
+		m_callbacks(std::move(_callbacks))
+	{}
+
+	Slot const& top() const
+	{
+		yulAssert(!m_data->empty());
+		return m_data->back();
+	}
+
+	void swap(Depth const& _depth)
+	{
+		yulAssert(swapReachable(_depth), "Stack too deep");
+		std::swap((*m_data)[depthToOffset(_depth).value], m_data->back());
+		if constexpr (!std::is_same_v<Callbacks, NoOpStackManipulationCallbacks>)
+			m_callbacks.swap(_depth.value);
+	}
+	void swap(Offset const& _offset) { swap(offsetToDepth(_offset)); }
```

Comment:
Wouldn't it be easier to do this the other way around? Implement `swap` with `Depth` using `swap` with `Offset`?

Reply:
I now see your point clearly. I was not thinking about the other stuff happening around the actual operation.
So I am fine with whatever you decide is the best here!